File: mobile_app/lib/data/models/lesson.dart
import 'package:json_annotation/json_annotation.dart';
import 'theme.dart';

part 'lesson.g.dart';

/// Book Author nested model
@JsonSerializable()
class BookAuthorNested {
  final int id;
  final String name;

  BookAuthorNested({
    required this.id,
    required this.name,
  });

  factory BookAuthorNested.fromJson(Map<String, dynamic> json) =>
      _$BookAuthorNestedFromJson(json);

  Map<String, dynamic> toJson() => _$BookAuthorNestedToJson(this);
}

/// Book nested model
@JsonSerializable()
class BookNested {
  final int id;
  final String name;
  final BookAuthorNested? author;

  BookNested({
    required this.id,
    required this.name,
    this.author,
  });

  factory BookNested.fromJson(Map<String, dynamic> json) =>
      _$BookNestedFromJson(json);

  Map<String, dynamic> toJson() => _$BookNestedToJson(this);
}

/// Teacher nested model
@JsonSerializable()
class TeacherNested {
  final int id;
  final String name;

  TeacherNested({
    required this.id,
    required this.name,
  });

  factory TeacherNested.fromJson(Map<String, dynamic> json) =>
      _$TeacherNestedFromJson(json);

  Map<String, dynamic> toJson() => _$TeacherNestedToJson(this);
}

/// Lesson Series model
@JsonSerializable()
class LessonSeries {
  final int id;
  final String name;
  final int year;
  @JsonKey(name: 'display_name')
  final String displayName;
  @JsonKey(name: 'teacher_id')
  final int teacherId;
  @JsonKey(name: 'book_id')
  final int? bookId;
  @JsonKey(name: 'theme_id')
  final int? themeId;
  @JsonKey(name: 'is_active')
  final bool isActive;
  @JsonKey(name: 'lessons_count')
  final int? lessonsCount;
  final TeacherNested? teacher;
  final BookNested? book;
  final AppThemeModel? theme;

  LessonSeries({
    required this.id,
    required this.name,
    required this.year,
    required this.displayName,
    required this.teacherId,
    this.bookId,
    this.themeId,
    required this.isActive,
    this.lessonsCount,
    this.teacher,
    this.book,
    this.theme,
  });

  factory LessonSeries.fromJson(Map<String, dynamic> json) =>
      _$LessonSeriesFromJson(json);

  Map<String, dynamic> toJson() => _$LessonSeriesToJson(this);
}

/// Lesson Series nested (simplified)
@JsonSerializable()
class LessonSeriesNested {
  final int id;
  final String name;
  final int year;
  @JsonKey(name: 'display_name')
  final String displayName;

  LessonSeriesNested({
    required this.id,
    required this.name,
    required this.year,
    required this.displayName,
  });

  factory LessonSeriesNested.fromJson(Map<String, dynamic> json) =>
      _$LessonSeriesNestedFromJson(json);

  Map<String, dynamic> toJson() => _$LessonSeriesNestedToJson(this);
}

/// Theme nested (simplified)
@JsonSerializable()
class ThemeNested {
  final int id;
  final String name;

  ThemeNested({
    required this.id,
    required this.name,
  });

  factory ThemeNested.fromJson(Map<String, dynamic> json) =>
      _$ThemeNestedFromJson(json);

  Map<String, dynamic> toJson() => _$ThemeNestedToJson(this);
}

/// Lesson nested (simplified) - for test questions
@JsonSerializable()
class LessonNested {
  final int id;
  final String? title;
  @JsonKey(name: 'lesson_number')
  final int lessonNumber;
  @JsonKey(name: 'display_title')
  final String? displayTitle;

  LessonNested({
    required this.id,
    this.title,
    required this.lessonNumber,
    this.displayTitle,
  });

  factory LessonNested.fromJson(Map<String, dynamic> json) =>
      _$LessonNestedFromJson(json);

  Map<String, dynamic> toJson() => _$LessonNestedToJson(this);
}

/// Lesson model
@JsonSerializable()
class Lesson {
  final int id;
  final String? title;
  @JsonKey(name: 'display_title')
  final String? displayTitle;
  @JsonKey(name: 'lesson_number')
  final int lessonNumber;
  @JsonKey(name: 'duration_seconds')
  final int? durationSeconds;
  @JsonKey(name: 'formatted_duration')
  final String? formattedDuration;
  @JsonKey(name: 'audio_url')
  final String? audioUrl;
  @JsonKey(name: 'audio_file_path')
  final String? audioFilePath;
  final String? description;
  final String? tags;
  @JsonKey(name: 'tags_list')
  final List<String>? tagsList;
  @JsonKey(name: 'is_active')
  final bool? isActive;
  @JsonKey(name: 'series_id')
  final int? seriesId;
  @JsonKey(name: 'teacher_id')
  final int? teacherId;
  @JsonKey(name: 'book_id')
  final int? bookId;
  @JsonKey(name: 'theme_id')
  final int? themeId;
  final LessonSeriesNested? series;
  final TeacherNested? teacher;
  final BookNested? book;
  final ThemeNested? theme;

  Lesson({
    required this.id,
    this.title,
    this.displayTitle,
    required this.lessonNumber,
    this.durationSeconds,
    this.formattedDuration,
    this.audioUrl,
    this.audioFilePath,
    this.description,
    this.tags,
    this.tagsList,
    this.isActive,
    this.seriesId,
    this.teacherId,
    this.bookId,
    this.themeId,
    this.series,
    this.teacher,
    this.book,
    this.theme,
  });

  factory Lesson.fromJson(Map<String, dynamic> json) =>
      _$LessonFromJson(json);

  Map<String, dynamic> toJson() => _$LessonToJson(this);
}

File: backend/app/schemas/lesson.py
"""
Pydantic schemas for Teacher, Series, Lesson models.
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field


# Lesson Teacher schemas
class LessonTeacherBase(BaseModel):
    """Base lesson teacher schema."""
    name: str = Field(..., max_length=255)
    biography: Optional[str] = None
    is_active: bool = True


class LessonTeacherCreate(LessonTeacherBase):
    """Schema for creating a lesson teacher."""
    pass


class LessonTeacherUpdate(BaseModel):
    """Schema for updating a lesson teacher."""
    name: Optional[str] = Field(None, max_length=255)
    biography: Optional[str] = None
    is_active: Optional[bool] = None


class LessonTeacherResponse(LessonTeacherBase):
    """Lesson teacher response schema."""
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class LessonTeacherWithCounts(LessonTeacherResponse):
    """Teacher with series and lessons count."""
    series_count: Optional[int] = 0
    lessons_count: Optional[int] = 0


# Lesson Series schemas
class LessonSeriesBase(BaseModel):
    """Base lesson series schema."""
    name: str = Field(..., max_length=255)
    year: int
    description: Optional[str] = None
    is_completed: bool = False
    order: int = 0
    is_active: bool = True


class LessonSeriesCreate(BaseModel):
    """Schema for creating a lesson series."""
    name: str = Field(..., max_length=255)
    year: int
    description: Optional[str] = None
    teacher_id: int
    book_id: Optional[int] = None
    theme_id: Optional[int] = None
    is_completed: bool = False
    order: int = 0
    is_active: bool = True


class LessonSeriesUpdate(BaseModel):
    """Schema for updating a lesson series."""
    name: Optional[str] = Field(None, max_length=255)
    year: Optional[int] = None
    description: Optional[str] = None
    teacher_id: Optional[int] = None
    book_id: Optional[int] = None
    theme_id: Optional[int] = None
    is_completed: Optional[bool] = None
    order: Optional[int] = None
    is_active: Optional[bool] = None


class LessonSeriesResponse(LessonSeriesBase):
    """Lesson series response schema."""
    id: int
    teacher_id: int
    book_id: Optional[int] = None
    theme_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Nested schemas for relationships (to avoid circular imports)
class LessonSeriesNested(BaseModel):
    """Nested series info for lesson responses."""
    id: int
    name: str
    year: int
    display_name: Optional[str] = None

    class Config:
        from_attributes = True


class TeacherNested(BaseModel):
    """Nested teacher info."""
    id: int
    name: str

    class Config:
        from_attributes = True


class ThemeNested(BaseModel):
    """Nested theme info."""
    id: int
    name: str

    class Config:
        from_attributes = True


class BookNested(BaseModel):
    """Nested book info."""
    id: int
    name: str

    class Config:
        from_attributes = True


class LessonSeriesWithRelations(LessonSeriesResponse):
    """Series with teacher, book, theme info."""
    teacher: Optional[TeacherNested] = None
    book: Optional[BookNested] = None
    theme: Optional[ThemeNested] = None
    display_name: Optional[str] = None


class LessonSeriesWithCounts(LessonSeriesWithRelations):
    """Series with lessons count."""
    lessons_count: Optional[int] = 0
    total_duration: Optional[str] = None  # Formatted duration


# Lesson schemas
class LessonBase(BaseModel):
    """Base lesson schema."""
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    lesson_number: Optional[int] = None
    duration_seconds: Optional[int] = None
    tags: Optional[str] = None
    is_active: bool = True


class LessonCreate(BaseModel):
    """Schema for creating a lesson."""
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    lesson_number: Optional[int] = None
    duration_seconds: Optional[int] = None
    tags: Optional[str] = None
    series_id: int
    book_id: Optional[int] = None
    teacher_id: Optional[int] = None
    theme_id: Optional[int] = None
    is_active: bool = True


class LessonUpdate(BaseModel):
    """Schema for updating a lesson."""
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    audio_path: Optional[str] = Field(None, max_length=500)
    original_audio_path: Optional[str] = Field(None, max_length=500)
    lesson_number: Optional[int] = None
    duration_seconds: Optional[int] = None
    tags: Optional[str] = None
    waveform_data: Optional[str] = None  # JSON array of waveform amplitude values
    series_id: Optional[int] = None
    book_id: Optional[int] = None
    teacher_id: Optional[int] = None
    theme_id: Optional[int] = None
    is_active: Optional[bool] = None


class LessonResponse(LessonBase):
    """Lesson response schema."""
    id: int
    audio_path: Optional[str] = Field(None, serialization_alias='audio_file_path')
    original_audio_path: Optional[str] = None
    series_id: int
    book_id: Optional[int] = None
    teacher_id: Optional[int] = None
    theme_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
        populate_by_name = True


class LessonWithRelations(LessonResponse):
    """Lesson with all related info (for mobile app)."""
    display_title: Optional[str] = None  # e.g., "Урок 1"
    formatted_duration: Optional[str] = None  # e.g., "30м 15с"
    audio_url: Optional[str] = None  # API URL for audio streaming
    tags_list: Optional[List[str]] = None  # Parsed tags
    waveform_data: Optional[str] = None  # JSON array of waveform amplitude values

    # Related entities
    series: Optional[LessonSeriesNested] = None
    teacher: Optional[TeacherNested] = None
    book: Optional[BookNested] = None
    theme: Optional[ThemeNested] = None


class LessonListItem(BaseModel):
    """Simplified lesson for list views."""
    id: int
    lesson_number: Optional[int] = None
    display_title: str
    duration_seconds: Optional[int] = None
    formatted_duration: Optional[str] = None
    audio_url: Optional[str] = None
    waveform_data: Optional[str] = None  # JSON array of waveform amplitude values
    teacher: Optional[TeacherNested] = None
    book: Optional[BookNested] = None

    class Config:
        from_attributes = True
